packages/dotenv-webpack-plugin/src/index.js

```diff
@@ -0,0 +1,148 @@
+/** @typedef {import("schema-utils/declarations/validate").Schema} Schema */
+/** @typedef {import("webpack").Compiler} Compiler */
+
+class DotenvWebpackPlugin {
+  /**
+   * Dotenv Webpack Plugin
+   * @param {Config} options - Configuration options
+   */
+  constructor(options = {}) {
+    validate(/** @type {Schema} */ (schema), options || {}, {
+      name: "DotenvWebpackPlugin",
+      baseDataPath: "options",
+    });
+
+    const currentDirectory = process.cwd();
+
+    const {
+      envFiles = [
+        `${currentDirectory}/.env.example`, // loaded in all cases
+        `${currentDirectory}/.env`, // loaded in all cases
+        `${currentDirectory}/.env.local`, // loaded in all cases, ignored by git
+        `${currentDirectory}/.env.[mode]`, // only loaded in specified mode
+        `${currentDirectory}/.env.[mode].local`, // only loaded in specified mode, ignored by git
+      ],
+      prefixes = ["process.env.", "import.meta.env."],
+      envVarPrefix = "PUBLIC_",
+    } = options;
+
+    this.options = {
+      envFiles,
+      prefixes,
+      envVarPrefix,
+    };
+  }
+
+  /**
+   * Check if file exists
+   * @param {Compiler} compiler - Webpack compiler
+   * @param {string} environmentFile - Path to environment file
+   */
+  fileExists(compiler, environmentFile) {
+    return new Promise((resolve) => {
+      const fs = compiler.inputFileSystem;
+
+      fs.stat(environmentFile, (err) => {
+        // File does not exist
+        if (err) {
+          return resolve(false);
+        }
+
+        return resolve(environmentFile);
+      });
+    });
+  }
+
+  /**
+   * Read file from path and parse it
+   * @param {Compiler} compiler - Webpack compiler
+   * @param {string} environmentFile - Path to environment file
+   */
+  readFile(compiler, environmentFile) {
+    return new Promise((resolve, reject) => {
+      const envVariables = {};
+
+      const fs = compiler.inputFileSystem;
+
+      this.fileExists(compiler, environmentFile).then((exists) => {
+        if (!exists) {
+          return resolve();
+        }
+
+        fs.readFile(environmentFile, (err, environmentFileContents) => {
+          if (err) {
+            const logger = compiler.getInfrastructureLogger("DotenvWebpackPlugin");
+            logger.error(`Could not read ${environmentFile}`);
+            return reject(err);
+          }
+
+          const parsedEnvVariables = dotenv.parse(environmentFileContents);
+          for (const [key, value] of Object.entries(parsedEnvVariables)) {
+            // only add variables starting with the provided prefix
+            if (key.startsWith(this.options.envVarPrefix)) {
+              for (let index = 0; index < this.options.prefixes.length; index++) {
+                const prefix = this.options.prefixes[index];
+                envVariables[`${prefix}${key}`] = value;
+              }
+            }
+          }
+
+          resolve(envVariables);
+        });
+      });
```

Comment:
Let's read all variables here and filter them above

Reply:
refactoring done